### PR TITLE
Offsite postgresql backups using WAL-E

### PR DESCRIPTION
**Pull request by surminus**
This PR strives to allow encrypted offsite backups to an S3 bucket, along with continuous archiving that also helps us with DR purposes.

Story: https://trello.com/c/sozQLBHM/42-spike-end-to-end-backup-procedure-for-bouncer

WAL-E: https://github.com/wal-e/wal-e

I have successfully tested this in Vagrant. The bucket in Integration has been deployed [from this change](https://github.com/alphagov/govuk-terraform-provisioning/pull/26).

Would love to hear some critical opinion on this!

### CODE REVIEW

File: modules/govuk_postgresql/manifests/wal_e/backup.pp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+#
+# [*wale_private_gpg_key*]
+#   The private GPG to import to allow encryption. If present, WAL-E will
+#   will encrypt the backups.
+#
+# [*wale_private_gpg_key_fingerprint*]
+#   The GPG key fingerprint for the above GPG key.
+#
+# [*hour*]
+#   The hour(s) the cron job runs.
```
Comment (mattbostock) [inline modules/govuk_postgresql/manifests/wal_e/backup.pp]: It'd be good to say that the cron job is for the daily base backup (to distinguish it from the incremental backups that happen continuously).

File: modules/govuk_postgresql/manifests/wal_e/backup.pp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+    file { '/usr/local/bin/wal-e_postgres_backup_push':
+      ensure  => present,
+      content => template('govuk_postgresql/usr/local/bin/wal-e_postgres_backup_push.erb'),
+      mode    => '0775',
+      require => Class['govuk_postgresql::wal_e::package'],
+    }
+
+    $threshold_secs = 28 * 3600
+    $service_desc = 'PostgreSQL WAL-E backup push'
```
Comment (mattbostock) [inline modules/govuk_postgresql/manifests/wal_e/backup.pp]: Should the service description mention 'base' or 'full' backup?

File: modules/govuk_postgresql/manifests/wal_e/package.pp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+#
+class govuk_postgresql::wal_e::package {
+  package { 'wal-e':
+    ensure   => present,
+    provider => pip,
+    require  => Class['govuk_postgresql::server'],
+  }
+
+  $dependencies = [
+    'lzop',
```
Comment (mattbostock) [inline modules/govuk_postgresql/manifests/wal_e/package.pp]: Nice and clean code :+1:

File: modules/govuk_postgresql/templates/etc/cron.d/wal-e_postgres_backup_push_cron.erb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+MAILTO='root'
```
Comment (mattbostock) [inline modules/govuk_postgresql/templates/etc/cron.d/wal-e_postgres_backup_push_cron.erb]: Do we need to send email? Could we have wal-e send its output to syslog?

**Comment by mattbostock** [comment]
I've made some minor comments but overall this looks excellent.

**Comment by surminus** [comment]
Thanks @mattbostock, updated as per your comments. Very much agree with your comment over syslogging, not sure whether we wanted it more verbose (added this as a separate commit).

**Comment by mattbostock** [comment]
@surminus Did you mean to squash 4109503?

I noticed you added logging, but we're not logging the output of the WAL-E script. There's an [example](https://github.com/alphagov/govuk-puppet/blob/321d0449b2e582252c080649c93650d2d1146a2f/modules/govuk/templates/usr/local/bin/es-rotate-passive-check.erb#L4) where we've done this elsewhere.

**Comment by surminus** [comment]
@mattbostock thanks, I've updated the logging; I thought the postgres log was verbose enough but there's no harm in pushing to syslog either. 

I also note I haven't added any tests which would be good to do.

File: modules/govuk_postgresql/templates/usr/local/bin/wal-e_postgres_backup_push.erb [hunk 5]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+set -e
+
+# Redirect stdout and stderr to syslog
+# exec 1> >(/usr/bin/logger -s -t $(basename $0)) 2>&1
```
Comment (mattbostock) [inline modules/govuk_postgresql/templates/usr/local/bin/wal-e_postgres_backup_push.erb]: Did you intend to leave this commented out?

**Comment by mattbostock** [comment]
:+1: LGTM

**Comment by mattbostock** [comment]
:+1:

**Comment by mattbostock** [comment]
:eggplant:

**Pull request merged**